File: backend/src/routes/pdf-routes.ts
import express, { Request, Response, NextFunction } from 'express';
import multer from 'multer';
import { PDFService } from '../services/pdf-service';

const router = express.Router();
const pdfService = new PDFService();

// Multer-Konfiguration für den Datei-Upload
const upload = multer({
  storage: multer.memoryStorage(),
  limits: {
    fileSize: 10 * 1024 * 1024, // 10MB Limit
  },
  fileFilter: (req: Request, file: Express.Multer.File, cb: multer.FileFilterCallback) => {
    console.log('Empfangene Datei:', file.originalname, 'Typ:', file.mimetype);
    if (file.mimetype === 'application/pdf') {
      cb(null, true);
    } else {
      cb(new Error('Nur PDF-Dateien sind erlaubt'));
    }
  },
});

// Error Handler für Multer
const handleMulterError = (err: any, req: Request, res: Response, next: NextFunction) => {
  if (err instanceof multer.MulterError) {
    if (err.code === 'LIMIT_FILE_SIZE') {
      return res.status(400).json({ error: 'Datei ist zu groß (Max: 10MB)' });
    }
    return res.status(400).json({ error: err.message });
  }
  if (err) {
    return res.status(400).json({ error: err.message });
  }
  next();
};

// Route zum Extrahieren von Text aus einer PDF
router.post(
  '/extract-text',
  (req: Request, res: Response, next: NextFunction) => {
    console.log('PDF-Route aufgerufen');
    next();
  },
  upload.single('file'),
  handleMulterError,
  async (req: Request, res: Response) => {
    try {
      console.log('PDF-Upload-Anfrage empfangen');
      
      if (!req.file) {
        console.log('Keine Datei im Request gefunden');
        return res.status(400).json({ error: 'Keine Datei hochgeladen' });
      }

      console.log('Verarbeite PDF:', req.file.originalname);
      const text = await pdfService.extractText(req.file.buffer);
      
      console.log('PDF erfolgreich verarbeitet, sende Antwort');
      res.json({ text });
    } catch (error) {
      console.error('Fehler bei der PDF-Verarbeitung:', error);
      res.status(500).json({ 
        error: error instanceof Error ? error.message : 'Fehler bei der PDF-Verarbeitung' 
      });
    }
  }
);

export default router; 

File: backend/routes/pdf_routes.py
from flask import Blueprint, request, jsonify
from werkzeug.utils import secure_filename
import os
from services.pdf_extractor import PDFExtractor
import logging

pdf_bp = Blueprint('pdf', __name__)
logger = logging.getLogger(__name__)

# Konfiguration
UPLOAD_FOLDER = 'uploads'
ALLOWED_EXTENSIONS = {'pdf'}

def allowed_file(filename):
    return '.' in filename and filename.rsplit('.', 1)[1].lower() in ALLOWED_EXTENSIONS

@pdf_bp.route('/extract', methods=['POST'])
def extract_pdf():
    """
    Einfache Route zum Testen der PDF-Extraktion.
    Erwartet eine PDF-Datei im Formular-Feld 'file'.
    """
    try:
        # Überprüfe ob eine Datei gesendet wurde
        if 'file' not in request.files:
            return jsonify({'error': 'Keine Datei gefunden'}), 400
            
        file = request.files['file']
        if file.filename == '':
            return jsonify({'error': 'Kein Dateiname'}), 400
            
        if not allowed_file(file.filename):
            return jsonify({'error': 'Nur PDF-Dateien erlaubt'}), 400
            
        # Sichere die Datei
        filename = secure_filename(file.filename)
        filepath = os.path.join(UPLOAD_FOLDER, filename)
        
        # Stelle sicher dass der Upload-Ordner existiert
        os.makedirs(UPLOAD_FOLDER, exist_ok=True)
        
        # Speichere die Datei
        file.save(filepath)
        
        # Extrahiere Text
        extractor = PDFExtractor()
        if not extractor.is_valid_pdf(filepath):
            os.remove(filepath)
            return jsonify({'error': 'Ungültige PDF-Datei'}), 400
            
        text = extractor.extract_text(filepath)
        
        # Lösche die temporäre Datei
        os.remove(filepath)
        
        if text is None:
            return jsonify({'error': 'Konnte keinen Text aus der PDF extrahieren'}), 400
            
        return jsonify({
            'success': True,
            'text': text,
            'length': len(text)
        })
        
    except Exception as e:
        logger.error(f"Fehler bei der PDF-Verarbeitung: {str(e)}")
        return jsonify({'error': str(e)}), 500 